lib/puma/server.rb

```diff
@@ -233,6 +233,11 @@ def pool_capacity
       @thread_pool&.pool_capacity
     end
 
+    # @!attribute [r] busy_threads
+    def busy_threads
+      @thread_pool&.busy_threads
```

Comment:
This call locks the mutex:

```
    # @!attribute [r] busy_threads
    # @version 5.0.0
    def busy_threads
      with_mutex { @spawned - @waiting + @todo.size }
    end
```

I'm wondering (aloud) if we can reduce the requirement that this be 100% accurate at the sake of reducing locking contention. As-is, there is a race condition in the other metrics we're collecting.

Reply:
do you mean, don't lock only the case of stats collection? or in all cases?

## if you mean don't lock only for stats collection

is that to avoid adding overhead caused by stats collection? i'm pretty sure it's not relevant, since this method is also frequently called in business logic (see bellow). but yeah, sacrificing 100% accurate is definitely acceptable for this stat AFAIK

## if you mean don't lock ever

busy_threads is used to make business logic decisions about work scheduling elsewhere in the code, so i believe in that case the mutex is necessary

Reply:
I've also wondered about the locking code in ThreadPool.  Changing some of the variables to Queues might allow some simplification.  That work was interrupted.

At present, `busy_threads` is used in:

https://github.com/puma/puma/blob/a59afabe782ec8869bb709ca991b32c0b2cc95fd/lib/puma/request.rb#L169

With the recent focus on keep-alive long-tail issues, that may be removed.  In the meantime, maybe it's best to leave the locking place?

Reply:
https://github.com/puma/puma/pull/3527